[DOCS] path_hierarchy tokenizer examples

geekpete opened the pull request:
Additional examples for a common use case for `path_hierarchy` tokenizer.

Fixes https://github.com/elastic/elasticsearch/issues/17138

elasticmachine commented:
Pinging @elastic/es-search

javanna commented:
@debadair @lcawl could you please review this? I would love to get it in if it looks good.

mayya-sharipova commented on docs/reference/analysis/tokenizers/pathhierarchy-tokenizer-examples.asciidoc:
```diff
@@ -0,0 +1,195 @@
+        "custom_hierarchy": {
+          "type": "path_hierarchy",
+          "delimiter": "/",
+          "search_analyzer": "keyword"
```
The first time I see `search_analyzer` parameter under `tokenizer`. Did you intend to put it under `mappings` for corresponding fields?

geekpete replied:
I vaguely remember seeing something like this before where the nesting was wrong but wasn't caught and worked anyway, let me verify I'm doing this correctly and if we need a bug ticket for extra validation to prevent the "wrong" way perhaps.

geekpete replied:
Yeah I think I've encountered this before perhaps, it appears to be ignored, you can put any text you want in there and it doesn't appear to validate:
```
  "settings": {
    "analysis": {
      "analyzer": {
        "custom_path_tree": {
          "tokenizer": "custom_hierarchy"
        },
        "custom_path_tree_reversed": {
          "tokenizer": "custom_hierarchy_reversed"
        }
      },
      "tokenizer": {
        "custom_hierarchy": {
          "type": "path_hierarchy",
          "delimiter": "/",
          "jimbob": "keyword"
        },
        "custom_hierarchy_reversed": {
          "type": "path_hierarchy",
          "delimiter": "/",
          "any_text_you_want": "keyword",
          "reverse": "true"
        }
      }
    }
  }
```

I'll log (or locate if I've already logged it) a bug ticket to address this.
For now, I'll take it out and nothing should functionally change and tests should work the same way.

mayya-sharipova commented on docs/reference/analysis/tokenizers/pathhierarchy-tokenizer-examples.asciidoc:
```diff
@@ -0,0 +1,195 @@
+
+[source,js]
+--------------------------------------------------
+PUT /file-path-test/
```
should be have just `file-path-test`  instead of `/file-path-test/` for consistency? In this and following requests.

geekpete commented:
This is a great question as to why this is a separate page.
The `TESTSETUP` block must come at the top of the file.
In the original Path Hierarchy Tokenizer page, there was already a `TESTSETUP` block so I couldn't add my secondary one below it without tests failing.

I considered I had two options at this point.
One to add my `TESTSETUP` details to the existing block but hiding it from display when rendered, probably also requiring additional section to show the original TESTSETUP steps in the docs but not execute tests on them.

Alternatively to take my entire detailed example to another page of it's own  to avoid the `TESTSETUP` clash.

I'm keen to see if there's a way to both keep the test logic simple and obvious and also include it on the original page though. I wondered if we might have some way to independently test the pages using the separate TESTSETUP blocks then join the pages together after those tests were done to form one page perhaps?

geekpete commented:
- Fixed the consistency issue with the GET/PUT
- Found a minor typo in a word and fixed
- Removed the invalid yet not validated `search_analyzer` parameters from the tokenizer blocks
  - logged a bug ticket for the validation https://github.com/elastic/elasticsearch/issues/41813
- anyone who goes to cherry pick this for v7+ will need to make some minor adjustments for types removal

mayya-sharipova commented:
@geekpete  Thanks for addressing the changes. The PR looks good to me. The only thing we need to do is to include this page somewhere. I am fine for it to be a separate page, but than we need to include a link to it in `pathhierarchy-tokenizer.asciidoc`,  adding something like this at the end of the page ( I am not completely sure about the syntax, we can ask the doc team to double check):

```
[float]
=== Detailed Example
See an  <<analysis-pathhierarchy-tokenizer-examples, example here>>:: 

include::tokenizers/pathhierarchy-tokenizer-examples.asciidoc[]
````

geekpete commented:
I'll aim to finish this today.

jrodewig commented on docs/reference/analysis/tokenizers/pathhierarchy-tokenizer.asciidoc:
```diff
@@ -170,3 +170,9 @@ If we were to set `reverse` to `true`, it would produce the following:
+=== Detailed Examples
+See detailed  <<analysis-pathhierarchy-tokenizer-examples, examples here>>:: 
+
+include::tokenizers/pathhierarchy-tokenizer-examples.asciidoc[]
```
```suggestion

```

I believe this `include::` statement is causing the build to fail. The path is relative here so you'd need to remove `tokenizers` for it to work.

However, you already link to the examples so including it here is a bit redundant. I suggest removing it all together.

jrodewig commented on docs/reference/analysis/tokenizers/pathhierarchy-tokenizer.asciidoc:
```diff
@@ -170,3 +170,9 @@ If we were to set `reverse` to `true`, it would produce the following:
+
+[float]
+=== Detailed Examples
+See detailed  <<analysis-pathhierarchy-tokenizer-examples, examples here>>:: 
```
```suggestion
See detailed  <<analysis-pathhierarchy-tokenizer-examples, examples here>>.
```

kat257 commented:
LGTM.
